src/cli/mod.ts

```diff
@@ -96,6 +102,81 @@ export function dzx() {
       }),
     );
 
+  async function moduleFromMarkdown(md: string) {
+    let mdRealPath;
+    let mdContent;
+
+    if (
+      md.startsWith("http://") ||
+      md.startsWith("https://")
+    ) {
+      mdContent = await fetch(md).then((r) => r.text());
+      mdRealPath = await Deno.makeTempFile({
+        prefix: "dzx_",
+        suffix: `_source.md`,
+      });
+
+      await Deno.writeTextFile(mdRealPath, mdContent);
+
+      console.log($.dim(`Markdown source saved to ${mdRealPath}`));
+    }
+
+    mdRealPath ??= await Deno.realPath(md);
+    mdContent ??= await Deno.readTextFile(mdRealPath);
+    const mdFileURL = path.toFileUrl(await Deno.realPath(mdRealPath));
+    const mdTokens = tokens(mdContent);
+
+    const validLangs = ["js", "javascript", "ts", "typescript"];
+    const codeContent: string[] = [];
+
+    mdTokens.forEach((token, idx) => {
+      if (
+        token.type === "start" && token.tag === "codeBlock" &&
+        token.kind === "fenced" &&
+        validLangs
+          .includes(token.language)
+      ) {
+        let cursor = idx + 1;
+
+        while (mdTokens.at(cursor)?.type === "text") {
+          const code = mdTokens.at(cursor);
+
+          // Silly typescript, can't narrow this down based on
+          // the while loop condition...ah well
+          if (code?.type === "text") {
+            codeContent.push(code.content);
+          }
+
+          cursor++;
+        }
+      }
+    });
```

Comment:
stupid idea from me. this will only work with simple code blocks. should be fine as it is.

Reply:
maybe we can write our own parse like zx it does but for now i'm also fine with the 3rd party lib.

Reply:
Hm, I can test this and see if it seems to work for all cases. The rules I landed on about which code blocks get executed and which don't are pretty rigid, so a regex may be sufficient here, but the tricky thing with Markdown is that it's super permissive and you can nest things like crazy. For instance, I don't think that regex will work for the following valid markdown:

```
# a header

> A blockquoted line
>
> ```ts
>   // nested code in a code block which when extracted with ^that regex, would have leading > characters I think?
> ```
```

I can test it out though, it would be nice not to rely on any third party modules (that you didn't also write) :+1:

Reply:
Ok, so I just did some testing and that regex, when run against the example file, produces the following script:

```
$.verbose = true;

const name = "World";
await $`echo "Hello ${name}!"`;// Note that we're using the `name`
// defined in the first code block
await $`echo "Hello again ${name}!"`;  // TLDR: keep your usage of `"file:///home/code/dzx/examples/markdown.md"` simple and naive-string-replacement friendly ❤️
  const importMetaUrl = "file:///home/code/dzx/examples/markdown.md";
  // this line will be replace at compile-time with (for example):
  // const importMetaUrl = "file:///home/user/code/example.md";
  console.log({ importMetaUrl });
  ```

- If you installed `dzx` without the `--no-check` flag, then your code blocks
  will be type checked. If you don't want that, you can either re-install `dzx`
  with the `--no-check` flag included, or add `// @ts-nocheck <reason>` as a
  fenced code block of its own at the top of your markdown file (or just before
  the first line of code in whatever your first fenced code block is).

- You do not need to add a `shebang` (e.g. `#!/usr/bin/env dzx`) to your code
  blocks, nor do you need to add a reference to the `dzx` types. Both of these
  will be automatically inserted at the top of the compiled (temp file)
  typescript module.

## Wrapping up

If you run _this Markdown file_ with `dzx`, the compiled script file will be
something like this:
```

Which is obviously looking good initially, but then it's a bit too greedy at the end. I think I could definitely spend some time iterating on the regex to get it more right, but I wonder if it's worth it to save the dependency which has already proven to parse these more complex blocks correctly? Thoughts?

Reply:
Actually, I am doing some looking around the internet to see if anyone has solved this regex so I don't have to iterate on it as much, and I'm finding some good stuff...I'll take another pass at this before re-requesting review.

Reply:
@c4spar Ok, I have what I think is a conclusion. I did some testing with various options I found online, and the one with the best results came from https://github.com/regexhq/gfm-code-block-regex which provides the regex:

```
/^(([ \t]*`{3,4})([^\n]*)([\s\S]+?)(^[ \t]*\2))/gm
```

This does great for most blocks, but can't handle blockquouted code and a few other more "advanced" scenarios that markdown permits. In my testing, the current rust based lib does handle these perfectly. So, in conclusion, I think we could go with the regex and drop the library, but it would require adding the stipulation that the supported markdown is more simple (probably ok honestly) or we can keep the lib and it supports more valid markdown syntax, but at the cost of having a dep. I am happy to go with either option, so I'll leave the final call up to you :)

Reply:
yeah a regex won't work in all cases, so i'm totally fine to add the dependency. we can always build our own parse later.